server/core/new_region_cache.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2020 PingCAP, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+type regionQueueNode struct {
+	region *RegionInfo
+	pre    *regionQueueNode
+	next   *regionQueueNode
+}
+
+type regionQueue struct {
+	start *regionQueueNode
+	end   *regionQueueNode
+	len   int
+}
+
+func newRegionQueue() *regionQueue {
+	return &regionQueue{
+		start: nil,
+		end:   nil,
+		len:   0,
+	}
+}
+
+func (queue *regionQueue) length() int {
+	return queue.len
+}
+
+// getRegions gets all RegionInfo from regionQueue.
+func (queue *regionQueue) getRegions() []*RegionInfo {
+	var regions []*RegionInfo
+	temp := queue.start
+	for temp != nil {
+		regions = append(regions, temp.region)
+		temp = temp.next
+	}
+	return regions
+}
+
+// findNode will find the regionQueueNode which has the region with regionID.
+func (queue *regionQueue) findNode(regionID uint64) *regionQueueNode {
+	temp := queue.start
+	for temp != nil {
+		if temp.region.GetID() == regionID {
+			return temp
+		}
+		temp = temp.next
+	}
+	return nil
+}
```

Comment:
This can be done faster using map?

Reply:
Map is a key-value store. It should be faster than scaning a linked list.  It's better to use map here. I will modify it.